tests/ui/type-alias-impl-trait/lazy_subtyping_of_opaques.rs

```diff
@@ -0,0 +1,31 @@
+#![feature(type_alias_impl_trait)]
+
+//! This test used to ICE rust-lang/rust#124891
+//! because we added an assertion for catching cases where opaque types get
+//! registered during the processing of subtyping predicates.
+
+type Tait = impl FnOnce() -> ();
+
+fn reify_as_tait() -> Thunk<Tait> {
+    //~^ ERROR expected a `FnOnce()` closure, found `()`
+    Thunk::new(|cont| cont /* opaque type gets constrained to `()`` here */)
```

Comment:
I assumed subtyping was happening here, because when checking whether `Thunk::new` can be called with this closure argument, we had to prove that the closure implements `ContFn`, and that somehow had to crate an obligation instead of just subtyping immediately. I can start logging things, but debugging this is going to be a pain.